### DISCUSSION TITLE
Managing Attributes in the Button Widget Actions

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#8271] [Q&A] [answered]
1. Open https://tiddlywiki.com/prerelease/
2. Create a new tiddler with below script

```
\procedure doTest(a) 
<$action-setfield $tiddler=Test $filed=text $value=<<a>> /> 
\end

<$button actions=<<doTest 100>> >
Button 1: {{Test}}
</$button>

<$button>
<<doTest 200>>
Button 2: {{Test}}
</$button>
```

Button 1 uses actions attributes and set the value of text field of Test tiddler to 100
Button 2 uses in-body actions and set the value of text field of Test tiddler to 200

Why button 1 does not work? Look at the header: `<$button actions=<<doTest 100>> >`
Why my action procedure does not handle `a` correctly?

**Comment by pmario** [comment]
Because, in the first procedure call `<$button actions=<<doTest 100>> >` the proc `<<doTest 100>>` is not evaluated. Wikitext / TiddlyWiki is a "text substitution engine" and not a programming language.

Calling `<$button actions=<<doTest 100>> >` is the same thing as if you call it like this: 

```
<$button actions="""<$action-setfield $tiddler=Test $filed=text $value=<<a>> />"""  >
```

So **it does not know about the parameter 100**

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/3bcc95fa-0d0b-48d1-8971-af5cf15b010a)

------

If you call it as follows, it works. 

```
<$button actions="""<<doTest 100>>""" >
```

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/927b70c9-29c3-4478-a797-6a22b5e30d11)

**Comment by pmario** [comment]
Check this out

```
\procedure variable() 123

\procedure doTest(a) 
<$action-setfield $tiddler=Test $filed=text $value=<<a>> /> 
\end

\procedure parameterVariable()
<$macrocall $name=doTest a=<<variable>> />
\end

<$button actions=<<parameterVariable>> >
Button 1: {{Test}}
</$button>

```

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/ea87861b-b187-4d62-ae25-632b5bac7f60)

---------

**Optimized**

```
\procedure variable() 123

\procedure doTest() 
<$action-setfield $tiddler=Test $filed=text $value=<<variable>> /> 
\end

<$button actions=<<doTest>> >
Button 1: {{Test}}
</$button>
```

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/3f657f2b-fc07-42b6-8f6d-707697af99f7)

Hope that makes sense

**Reply by kookma** [reply]
Thank you, Mario, for your detailed explanation. I understand the magic behind """<<doTest 100>>""".

**Comment by Jermolene** [comment, marked as answer]
Hi @kookma

> Why button 1 does not work? 

The most concise explanation is that the "actions" attribute is being set to the value of a procedure. But procedures are not wikified in this situation, instead the plain text of the procedure is used as the attribute value. The parameters of the procedure will be ignored too.
